### ISSUE TITLE
General feedback does not update in quiz already taken

### ISSUE THREAD

**Issue by alcarola** [maths/moodle-qtype_stack#1119]
Hi, I again organized an electronic exam using STACK a couple of weeks ago. I later fixed a mistake in my model solution and I think that the grading was corrected using an automatic regrade.

Now my problem is that the students still see the initial model solution in the general feedback, and not my corrected version. A student now spent considerable time on reporting that the old model solution was wrong, and me then had to spend time on sorting this out. (The grading of that particular student was at least correct.) The quiz is set to always use the latest version of the question.

I suppose it cannot be the intended behavior to display old model solutions instead of the most updated one? What information do I need to provide in order to help to find and correct the problem?

**Comment by aharjula** [comment]
Just to confirm, have you done the re-grade for all students? Even those that would not have had any issues requiring it? 

Doing re-grade should switch them to the new version, and that should then lead to the display of the new version's content, including the general feedback. If it does not, then I fear this might have something to do with Moodle itself.

**Comment by aharjula** [comment]
For those wondering, the quiz might be set to use the newest version, but the student's attempt will use the version that was newest when that attempt started. Until otherwise updated, and as far as I know re-grading is the only way to switch the version after that quiz attempt has started.

**Comment by alcarola** [comment]
Yes, doing a full regrade of the quiz seems to have fixed the problem! Many thanks!

Usually, I only do a regrade of those whose grades change, but in the sequel, I will have to try to remember this. 

Perhaps this is behavior that one would like to change in the future?

**Comment by aharjula** [comment]
Maybe not... This is a behaviour that makes sense as it ensures that the version of the question does not change from under the student unless someone actually takes action and does the re-grade. We just need to understand that this is how Moodle works now that it has those versioned questions and that this might be something desirable for various reasons. It just happens to require more actions, some of which we now need to learn to do, than it did before Moodle 4, but it does give some benefits for some edge cases and is somewhat cleaner in logic.

I must say that this has not been an easy thing to re-train users for. Most people do not see the logic of having to re-grade incomplete attempts to switch the question versions during a running quiz or otherwise, but that is what we have, and as long as there are no other means for doing that in the quiz module, this is what we have to do.

We just need to accept that we now act on a version level and that if the question has been graded using a particular version, it will then need to display the general/specific feedback of that version, until the active version changes. We really cannot mix versions and actually pick something from a version defined by the quiz instead of the attempt, that would lead to chaos.

**Comment by alcarola** [comment]
I agree with all that you say.

One idea would be to make the choice of question versions somehow more visible than it is now, for instance by adding an option somewhere in the quiz to either keep the version which was latest at the attempt, or to use the latest version. As a simple fix, this could be added to the menu where question version is selected, but perhaps there is also some better way.

**Comment by aharjula** [comment]
Indeed, however, for that to happen, someone needs to go to the Moodle people, in particular the quiz-module people, and make them believe in that. We cannot do that from the point of view of a single-question type.

**Comment by timhunt** [comment]
In the context of this, I want to mention:

The open university is working on an enhancement to the Moodle quiz regrade feature, so that you can just regrade one question in all attempts. Hopefully that will get into Moodle 4.4. (Sadly new features can only go into new major releases.)

(And, the main reason for doing this is for our STACK quizzes, where it should be a big performance win.)
